### ISSUE TITLE
Add option to set the default initialization options for a language server

### ISSUE THREAD

**Issue by jez** [autozimu/LanguageClient-neovim#1107]
## Is your feature request related to a problem? Please describe.

Specifying initialization options is repetitive.
Each individual project needs a new `.vim/settings.json` file.
Changes to the initialization options for a language server need require many `.vim/settings.json` files.

## Describe the solution you'd like

An option in Vim script to set the default initialization options for a language server.

```vim
let g:LanguageClient_defaultInitializationOptions = {}
let g:LanguageClient_defaultInitializationOptions.ruby = {
      \ "supportsOperationNotifications": v:true,
      \ }
```

In general, initialization options are overwhelmingly used to say "the LSP client supports this functionality."
That is, initialization options are frequently a function of the client (global) not each project.

## Describe alternatives you've considered

Currently `LanguageClient-neovim` reads from a JSON file (`.vim/settings.json`) in the current directory.
An alternative would be to also search for this file globally. That is undesirable:

- Vim plugins that configure LanguageClient-neovim on behalf of a user would not be able to control these settings.
- Vim is configured in Vim script. Vim users expect to use Vim script (not JSON) to configure their plugins.
- Using a variable lets the option be configured in code. For example, I currently have code to set different server options if my Ruby language server was built in release or debug mode (e.g., add some extra debug mode options for logging that are slow). Using code for initialization options would make it easier to test, for example, experimental / debug-only features.

**Comment by martskins** [comment]
To answer your "each individual project needs a new settings.json file" point, you can set the settings path to be used with `let g:LanguageClient_settingsPath = /your/path`, and it doesn't have to be within your project folder. So you could either set that to something like `let g:LanguageClient_settingsPath = ~/.vim/settings.json` or create a per-language config in your `ftplugin` folder and having a different setting for each language. That would essentially have the same effect as having a global config.

```vim
" contents of ~/.config/nvim/ftplugin/ruby.vim
let g:LanguageClient_settingsPath = ~/.vim/ruby/settings.json
```


```vim
" contents of ~/.config/nvim/ftplugin/rust.vim
let g:LanguageClient_settingsPath = ~/.vim/rust/settings.json
```

Your point about vimscript being preferable over JSON I think is valid, but I think ultimately most servers document the configuration with JSON examples so it's easier to use JSON.
Having said that, I'm not opposed to using vimscript for that, and I think there's value in adding a config value for this, but I think it would be weird to have this setting be a vimscript setting and the workspace local setting be a JSON one. Especially because we'd probably have to merge (or at least choose one of) the two settings.

So to summarise, I see the value in adding that, but it comes with a lot of added complexity as well, and the "global config" can already be configured so I'm not sure if it's worth going there.

**Comment by jez** [comment]
Ok.

I find it frustrating whenever I find Vim plugins that invent their own config formats instead of using Vim script (coc.nvim is guilty of the same sin). But it's your decision.

I have implemented what I want to based on `LanguageClient_settingsPath` here:

https://github.com/jez/vim-sorbet/blob/bec2698771f4392c2d0f75a4bad537c41beae707/ftplugin/ruby.vim#L13-L19

I would still appreciate this feature, but I have a workaround so if you want to close I'm fine with that too.

**Comment by martskins** [comment]
I agree with that feeling, but I reckon that this case was a little different, since those settings could be workspace settings, so adding them to the vim configuration was not really an option (or at least I can't think now how that would look). But more importantly, I think that supporting default initialisation options as vimscript and workspace level settings as JSON would be a little too inconsistent, and dropping support for JSON would break the plugin for many people.

Having said all that, this isn't really my decision neither, it's really up to @autozimu , so maybe he has a different opinion.

**Comment by martskins** [comment]
@jez I was playing with some ideas for this and ended up pushing some changes to a draft PR I had for something related. In that PR you can now specify default `initializationOptions` for a configured server in the .vimrc file. Workspace local settings will still need to be specified as they are now (with the settings.json file), but I reckon most cases would be covered by this.

See #1116

**Comment by martskins** [comment]
@jez Just merged #1116 in the `dev` branch.

That means that you can now specify your server command like this if you want to specify default initialization options for a server:

```
let g:LanguageClient_serverCommands = {
    \ 'ruby': {
    \     'name': 'sorbet',
    \     'command': ['sorbet'],
    \     'initializationOptions': {
    \         'supportsOperationNotifications': v:true
    \     }
    \ }
    \}
```

I'm closing this issue now that it's solved in dev and I'll look into cutting a release early next week if all works as expected. Feel free to re-open if it doesn't really solve what you had in mind though.
